configure.ac

```diff
@@ -676,33 +676,54 @@ if test x"${with_d}" = xno; then
   AC_MSG_NOTICE([Disabling D])
   D2COMPILER=
 else
-  old_ac_ext=$ac_ext
-  ac_ext=d
 
   if test -z "$D2COMPILERBIN" ; then
-    AC_CHECK_PROGS(D2COMPILER, dmd gdmd ldmd2 ldc2)
-
+    AC_CHECK_PROGS(D2COMPILER, dmd ldmd2 ldc2 gdmd gdc)
     if test -n "$D2COMPILER" ; then
+      old_ac_ext=$ac_ext
+      ac_ext=d
+      case $D2COMPILER in
+        gdc)
+          d_out=-oconftest.out
+          D2COMPILER_TYPE=GCC
+          ;;
+        ldc2|ldmd2)
+          d_out=-ofconftest.out
+          D2COMPILER_TYPE=LLVM
+          ;;
+        *)
+          d_out=-ofconftest.out
+          D2COMPILER_TYPE=DMD
+          ;;
+      esac
       AC_MSG_CHECKING(whether the D2 compiler works)
       cat > conftest.$ac_ext <<_ACEOF
 import std.algorithm;
 void main() {
 }
 _ACEOF
-      rm -f conftest.$ac_objext
+      rm -f conftest.out
       AS_IF(
-        [$D2COMPILER conftest.$ac_ext 2>&AS_MESSAGE_LOG_FD && test ! -s conftest.err && test -s conftest.$ac_objext],
+        [$D2COMPILER conftest.$ac_ext $d_out 2>&AS_MESSAGE_LOG_FD && test ! -s conftest.err && test -s conftest.out],
         [AC_MSG_RESULT([yes])],
         [_AC_MSG_LOG_CONFTEST AC_MSG_RESULT([no])
         D2COMPILER=]
       )
-      rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext
+      rm -f core conftest.err conftest.$ac_objext conftest.$ac_ext conftest.out
+      ac_ext=$old_ac_ext
     fi
   else
     D2COMPILER="$D2COMPILERBIN"
+    if echo $D2COMPILER | grep gdc >/dev/null 2>&1 ; then
+      D2COMPILER_TYPE=GCC
+    elif echo $D2COMPILER | grep ldc2 >/dev/null 2>&1 ; then
+      D2COMPILER_TYPE=LLVM
+    elif echo $D2COMPILER | grep ldmd2 >/dev/null 2>&1 ; then
+      D2COMPILER_TYPE=LLVM
+    else
+      D2COMPILER_TYPE=DMD
+    fi
```

Comment:
More efficient to use `case` to do this, since each `grep` has to fork() then exec an external command whereas `case` is a shell built-in:

```
case $D2COMPILER in
*gdc*) D2COMPILER_TYPE=GCC ;;
*ldc2*|*ldmd2*) D2COMPILER_TYPE=LLVM ;;
*) D2COMPILER_TYPE=DMD ;;
esac
```

Though maybe we don't need it at all if we use `dmd` (though I guess this also gets used for the testcase that's a problem with ldc - not got to that yet).

Reply:
I think we can use conditional compilation to address that testcase: https://dlang.org/spec/version.html

So inside the `runme` for d we'd have something like:

```
version(LDC) {
// Emit a message that we're skipping the test and why
} else {
// Existing code to run the test
}
```

Reply:
GCC uses `-o` while LLVM and DMD uses `-of` for output file name.
So it have to be in the configuration and Make files.
We can add `-version  $(D2COMPILER_TYPE)` to compilers.
For the `li_boost_shared_ptr` test.
I'll give it a try.

Reply:
I did not find any reference that explain what versions exist, apart for using `-version` command line.
As we already probe in the configure, we can simply forward the information to the compiler.

Reply:
The page I linked to above lists the pre-defined versions further down:

https://dlang.org/spec/version.html#predefined-versions

So `LDC` is already defined by `ldc` and there are predefined versions of other compilers too, so there's no need for us to define them.  That was why I suggested using this as it avoids needing per-compiler special cases.  I've pushed a commit which simplifies that.

> GCC uses -o while LLVM and DMD uses -of for output file name.
> So it have to be in the configuration and Make files.

Sorry, I didn't explain my point about gdc vs gdmd clearly I guess.  I'm saying we should just support GCC via the `gdmd` wrapper rather than trying to support using `gdc` too.  The whole point of `gdmd` is to allow using GCC to compiler D but with options compatible with `dmd`, so let's make use of that and then we can just call the D compiler with the same options without needing extra machinery.  I've pushed a commit to do that too.

Note this is only about what we support using in SWIG's testsuite and examples - we aren't imposing restrictions on how people compile the code SWIG generates for libraries they're wrapping.